support/k8s/spire-adm-webhook/webhook.go

```diff
@@ -0,0 +1,225 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"path"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/spire/pkg/common/idutil"
+	"github.com/spiffe/spire/proto/spire/api/registration"
+	"github.com/spiffe/spire/proto/spire/common"
+	"github.com/zeebo/errs"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	admv1beta1 "k8s.io/api/admission/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type WebhookConfig struct {
+	Log         logrus.FieldLogger
+	R           registration.RegistrationClient
+	TrustDomain string
+	Cluster     string
+	PodLabel    string
+}
+
+type Webhook struct {
+	c WebhookConfig
+}
+
+func NewWebhook(config WebhookConfig) *Webhook {
+	return &Webhook{
+		c: config,
+	}
+}
+
+func (w *Webhook) Initialize(ctx context.Context) error {
+	// ensure there is a node registration entry for PSAT nodes in the cluster.
+	return w.createEntry(ctx, &common.RegistrationEntry{
+		ParentId: idutil.ServerID(w.c.TrustDomain),
+		SpiffeId: w.nodeID(),
+		Selectors: []*common.Selector{
+			{Type: "k8s_psat", Value: fmt.Sprintf("cluster:%s", w.c.Cluster)},
+		},
+	})
+}
+
+func (w *Webhook) ReviewAdmission(ctx context.Context, req *admv1beta1.AdmissionRequest) (*admv1beta1.AdmissionResponse, error) {
+	w.c.Log.WithFields(logrus.Fields{
+		"namespace": req.Namespace,
+		"name":      req.Name,
+		"kind":      req.Kind.Kind,
+		"version":   req.Kind.Version,
+		"operation": req.Operation,
+	}).Debug("ReviewAdmission called")
+
+	if err := w.reviewAdmission(ctx, req); err != nil {
+		return nil, err
+	}
+
+	return &admv1beta1.AdmissionResponse{
+		UID:     req.UID,
+		Allowed: true,
+	}, nil
+}
+
+// reviewAdmission handles CREATE and DELETE requests for pods in
+// non-kubernetes namespaces. Ideally the ValidatingAdmissionWebhook
+// configuration has filters in place that prevent the webhook from being
+// called for requests other than these but just in case, we filter here as
+// well.
+func (w *Webhook) reviewAdmission(ctx context.Context, req *admv1beta1.AdmissionRequest) error {
+	switch req.Namespace {
+	case metav1.NamespacePublic, metav1.NamespaceSystem:
+		return nil
+	}
+
+	if req.Kind != (metav1.GroupVersionKind{Version: "v1", Kind: "Pod"}) {
+		return nil
+	}
+
+	switch req.Operation {
```

Comment:
Does it make sense for us to log or otherwise report that we are receiving webhooks for requests other than the ones we are interested in? If we are down, pods can't be scheduled or deleted... that seems pretty bad to begin with, but I wonder if there are other implications for non-CREATE/DELETE actions that we'd prefer not to interfere with in that case? We could let the user know that they are misconfigured

Reply:
yeah logging a warning sounds good.